### PR TITLE
Swarm GUI

### PR DESCRIPTION
**Pull request by stephendade**
This PR adds a swarm management GUI for managing and controlling swarms of vehicles.

Each swarm takes up a single row in the GUI, with the swarm leader being the leftmost vehicle in each row.

Any vehicle not in a swarm (or swarm leader not found) is put in the bottom "unassigned" row.

Features:
- Allows simultaneous control and monitoring of vehicles
- GUI will auto-generate based on detected vehicles
- Buttons for commonly-used commands on a per-vehicle and per-swarm basis
- Able to show multiple separate swarms
- Vehicles will go red if link is lost (no heartbeat for 4 seconds)

Requirements:
- Swarm must be pre-configured (FOLL_* params set)
- 1080P screen (or larger) recommended. The GUI does take up a fair bit of screen space for larger swarms

Screenshot from 2021-12-08 10-05-30

Clarification of the buttons:
- "Kill" sends a force-disarm. Button needs to be double-clicked to confirm, in order to prevent accidental commands
- "Guided T/O 10m" is a guided takeoff to 10m altitude. Altitude can be changed via ``swarm set takeoffalt n``
- The "All" button applies a command to all vehicles in swarm
- The "Get offsets" button will refresh all vehicle offset text labels
- The "Reset Layout" button will regenerate the entire GUI

Tested on Python2 and 3, Ubuntu20 and Windows.

To test in SITL, run ``./ardupilot/libraries/SITL/examples/follow-copter.sh`` and ``follow-mavproxy``, then type in ``module load swarm`` into MAVProxy.

@auturgy

### CODE REVIEW

**Comment by peterbarker** [comment]
Redraw issue - particularly on window resize:

image

File: MAVProxy/modules/mavproxy_swarm.py [hunk 1]
```diff
@@ -0,0 +1,953 @@
...
+    def __init__(self, state, title, params, takeoffalt):
+        self.state = state
+        wx.Frame.__init__(self, None, title=title, size=(
+            1400, 500), style=wx.DEFAULT_FRAME_STYLE)
+        self.panel = scrolled.ScrolledPanel(self)
+
+        # Icon
+        try:
+            self.SetIcon(icon.SimpleIcon("CONSOLE").get_ico())
+        except Exception:
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:497]: `Exception` should never be caught; surely there's some tighter exception you can catch?

File: MAVProxy/modules/mavproxy_swarm.py [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+                widget.updatetakeoffalt(alt)
+
+    def updateHB(self, vehHB):
+        '''Update the GUI panels if we've lost link to vehicle
+        Panel goes red if more than 4 sec since last HB'''
+        children = self.sizer.GetChildren()
+        # update panel
+        for child in children:
+            widget = child.GetWindow()
+            if isinstance(widget, vehiclePanel):
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:566]: an early-continue pattern here will save a bunch of indenting

File: MAVProxy/modules/mavproxy_swarm.py [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+        # Periodic event re-get params (0.1 Hz)
+        self.RerequestParams_timer = mavutil.periodic_event(0.1)
+
+        # List of any vehicles we still need to get params for
+        self.vehParamsToGet = []
+
+        # All vehicle positions. Dict. Key is sysid, value is tuple of (lat,lon,alt)
+        self.allVehPos = {}
+
+        self.validVehicles = [mavutil.mavlink.MAV_TYPE_FIXED_WING,
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:766]:
```suggestion
        self.validVehicles = frozenset([mavutil.mavlink.MAV_TYPE_FIXED_WING,
```

File: MAVProxy/modules/mavproxy_swarm.py [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+                            sysid, compid, parmString(parm), -1))
+
+    def mavlink_packet(self, m):
+        '''handle incoming mavlink packets'''
+        mtype = m.get_type()
+        sysid = m.get_srcSystem()
+        compid = m.get_srcComponent()
+
+        # add new vehicle on new heartbeat
+        if mtype == 'HEARTBEAT' and m.type in self.validVehicles:
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:917]: Early-continue instead of indented block here

File: MAVProxy/modules/mavproxy_swarm.py [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+            # add to GUI if vehicle not seen before
+            if not (sysid in [sysidList[0] for sysidList in self.vehicleListing] and compid in [compidList[1] for compidList in self.vehicleListing]):
+                self.vehicleListing.append((sysid, compid, 0, m.type))
+                # figure out leader for vehicle - check FOLL_SYSID
+                self.vehParamsToGet.append((sysid, compid))
+                self.needGUIupdate = True
+                self.vehicleLastHB[(sysid, compid)] = time.time()
+
+        # Only send these packets on if the vehicle is already in the list
+        if (sysid, compid) in self.vehicleLastHB.keys():
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:928]: Early-continue here too

File: MAVProxy/modules/mavproxy_swarm.py [hunk 6]
```diff
@@ -0,0 +1,953 @@
...
+        # add new vehicle on new heartbeat
+        if mtype == 'HEARTBEAT' and m.type in self.validVehicles:
+
+            # add to GUI if vehicle not seen before
+            if not (sysid in [sysidList[0] for sysidList in self.vehicleListing] and compid in [compidList[1] for compidList in self.vehicleListing]):
+                self.vehicleListing.append((sysid, compid, 0, m.type))
+                # figure out leader for vehicle - check FOLL_SYSID
+                self.vehParamsToGet.append((sysid, compid))
+                self.needGUIupdate = True
+                self.vehicleLastHB[(sysid, compid)] = time.time()
```
Comment (peterbarker) [inline MAVProxy/modules/mavproxy_swarm.py:925]: Redundant given the code below

**Review by peterbarker — CHANGES_REQUESTED** [review]
There seems to be a lot of pointless redrawing going on when not doing anything with the interface; the items seem to be blinking a bit as if they are being removed/added.

```
ERROR in command ['set']: list index out of range
Traceback (most recent call last):
  File "/home/pbarker/.local/lib/python3.6/site-packages/MAVProxy-1.8.46-py3.6.egg/EGG-INFO/scripts/mavproxy.py", line 761, in process_stdin
  File "/home/pbarker/.local/lib/python3.6/site-packages/MAVProxy-1.8.46-py3.6.egg/MAVProxy/modules/mavproxy_swarm.py", line 790, in cmd_swarm
    self.gui.changesetting(args[1], args[2])
```

Sadly the gui glitching is making it unusable for me ATM.  Ping me if you'd like to see what I'm seeing.  I don't think that's actually a blocker (pretty low threshold for merging modules in MAVProxy) - but if it isn't working for me it might not work for others too.

**Comment by stephendade** [comment]
>Sadly the gui glitching is making it unusable for me ATM. Ping me if you'd like to see what I'm seeing. I don't think that's actually a blocker (pretty low threshold for merging modules in MAVProxy) - but if it isn't working for me it might not work for others too.

Huh ... I've not come across that issue. What wxpython version and desktop manager are you using?

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looking much better in testing here now - the blinking is gone!

However:
 - mouse-overing (or un-mouse-overing) a button seems to cause all buttons to blink
 - mode changes are not reflected in the gui:
   - image
 - we're using the incorrect orientation for vehicles:
    - image
    - not your bug, probably
  - still getting this stuff: ```
Swarm: Got leader sysid 0 for vehicle 1
Swarm: Got leader sysid 0 for vehicle 3
Swarm: Got leader sysid 0 for vehicle 4
Swarm: Got leader sysid 0 for vehicle 5
Swarm: Got leader sysid 0 for vehicle 2
Swarm: Got leader sysid 0 for vehicle 1
Swarm: Got leader sysid 0 for vehicle 3
Swarm: Got leader sysid 0 for vehicle 4
Swarm: Got leader sysid 0 for vehicle 5
Swarm: Got leader sysid 0 for vehicle 2
```


I think the mode-change-needs-updating-in-interface is really the only thing that absolutely should be fixed before merge.

**Comment by stephendade** [comment]
> mode changes are not reflected in the gui: 

Fixed

> we're using the incorrect orientation for vehicles: 

Fixed

>still getting this stuff:

Fixed

I couldn't replicate the button flickering. If it's not too annoying, I'll leave it for the time being.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**